README.md
# SendGrid SMTP API Java Bindings

This module lets you build SendGrid's SMTP API headers with simplicity.

[![BuildStatus](https://travis-ci.org/revinate/sendgrid-smtpapi-java.svg?branch=master)](https://travis-ci.org/revinate/sendgrid-smtpapi-java)
[![BuildStatus](https://maven-badges.herokuapp.com/maven-central/com.revinate/sendgrid-smtpapi-java/badge.svg)](https://maven-badges.herokuapp.com/maven-central/com.revinate/sendgrid-smtpapi-java)

## Requirements

Java 1.6 and later.

## Installation

### Maven

Add this dependency to your project's POM:

```xml
<dependency>
  <groupId>com.revinate</groupId>
  <artifactId>sendgrid-smtpapi-java</artifactId>
  <version>2.0.0</version>
</dependency>
```

### Gradle

Add this dependency to your project's build script:

```groovy
compile 'com.revinate:sendgrid-smtpapi-java:2.0.0'
```

## Usage

### Create header

```java
import com.revinate.sendgrid.smtpapi.*;
SmtpApi header = new SmtpApiImpl();
```

### Get header value

```java
String headerValue = header.toSmtpApiHeader();
```

If you need the unescaped JSON string:

```java
String rawHeaderValue = header.toRawSmtpApiHeader();
```

### [To](http://sendgrid.com/docs/API_Reference/SMTP_API/index.html)

```java
header.addSmtpApiTo("email@email.com");
// or
header.addSmtpApiTo("email@email.com", "Email User");

List<String> tos = header.getSmtpApiTos();
```

### [Substitutions](http://sendgrid.com/docs/API_Reference/SMTP_API/substitution_tags.html)

```java
header.addValueToSubstitution("key", "value");

List<String> substitution = header.getSubstitution("key");
```

### [Unique Arguments](http://sendgrid.com/docs/API_Reference/SMTP_API/unique_arguments.html)

```java
header.setUniqueArg("key", "value");

String arg = header.getUniqueArg("key");
```

### [Categories](http://sendgrid.com/docs/API_Reference/SMTP_API/categories.html)

```java
header.addCategory("category");

List<String> categories = header.getCategories();
```

### [Sections](http://sendgrid.com/docs/API_Reference/SMTP_API/section_tags.html)

```java
header.setSection("key", "section");

String section = header.getSection("key");
```

### [Filters](http://sendgrid.com/docs/API_Reference/SMTP_API/apps.html)

```java
header.setSettingInFilter("filter", "setting", "value");
header.setSettingInFilter("filter", "setting", 1);

Map<String, Object> filter = header.getFilter("filter");
```

### [ASM Group ID](https://sendgrid.com/docs/User_Guide/advanced_suppression_manager.html)

```java
header.setAsmGroupId(1);

Integer groupId = header.getAsmGroupId();
```

### [Scheduling](https://sendgrid.com/docs/API_Reference/SMTP_API/scheduling_parameters.html)

```java
header.setSendAt(1416427645);

Integer sendAt = header.getSendAt();
```

### IP Pool
```java
header.setIpPool("transactional");

String ipPool = header.getIpPool();
```

## License

Licensed under the MIT License.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [2.0.0] - 2016-2-12
### Changed
- Do not expose JSON implementation to clients. Use collections instead of
arrays in interfaces. Rename methods where appropriate. This version is not
backwards-compatible with earlier ones.

## [1.2.1] - 2016-2-2
### Changed
- Renamed packages and artifact name. This breaks backwards compatibility.

### Added
- IP pool support

## [1.2.0] - 2015-5-14
### Added
- Support for raw unescaped JSON string - via [#10](https://github.com/sendgrid/smtpapi-java/pull/10)
- Version sanity tests


src/test/java/com/sendgrid/smtpapi/SMTPAPITest.java
package com.sendgrid.smtpapi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.skyscreamer.jsonassert.JSONAssert;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

import static org.junit.Assert.fail;

public class SMTPAPITest {

    SMTPAPI test;

    @Before
    public void initialize() {
        test = new SMTPAPI();
    }

    @Test
    public void testVersion() {
        Assert.assertEquals("1.2.1", test.getVersion());
    }

    @Test
    public void testBuildGradleVersion() {
        try {
            BufferedReader br = new BufferedReader(new FileReader("./build.gradle"));
            String line = br.readLine();
            String regex = "version\\s*=\\s*'" + test.getVersion() + "'";

            while (line != null) {
                if (line.matches(regex)) {
                    br.close();
                    return;
                }
                line = br.readLine();
            }
            br.close();
            fail("build.gradle version does not match");
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Test
    public void testAddTo() throws JSONException {
        test.addTo("john@doe.com");
        String[] expected = new String[]{"john@doe.com"};
        Assert.assertArrayEquals(expected, test.getTos());
    }

    @Test
    public void testAddTos() throws JSONException {
        String[] expected = new String[]{"john@doe.com"};
        test.addTos(expected);
        Assert.assertArrayEquals(expected, test.getTos());
    }

    @Test
    public void testSetTos() throws JSONException {
        String[] expected = new String[]{"john@doe.com", "doe@john.com"};
        test.setTos(expected);
        Assert.assertArrayEquals(expected, test.getTos());
    }

    @Test
    public void testAddSubstitution() throws JSONException {
        test.addSubstitution("-name-", "John");
        JSONObject testObject = test.getSubstitutions();
        JSONObject expected = new JSONObject().put("-name-", new JSONArray().put("John"));
        JSONAssert.assertEquals(expected, testObject, false);
    }

    @Test
    public void testAddSubstitutions() throws JSONException {
        test.addSubstitutions("-name-", new String[]{"John", "Doe"});
        JSONObject testObject = test.getSubstitutions();
        JSONObject expected = new JSONObject().put("-name-", new String[]{"John", "Doe"});
        JSONAssert.assertEquals(expected.toString(), testObject.toString(), false);
    }

    @Test
    public void testAddUniqueArg() throws JSONException {
        test.addUniqueArg("key", "value");
        JSONObject expected = new JSONObject().put("key", "value");
        JSONAssert.assertEquals(expected, test.getUniqueArgs(), false);
    }

    @Test
    public void testAddCategory() throws JSONException {
        test.addCategory("test");
        String[] expected = new String[]{"test"};
        Assert.assertArrayEquals(expected, test.getCategories());
    }

    @Test
    public void testAddCategoryUnicode() throws JSONException {
        test.addCategory("カテゴリUñicode");
        test.addCategory("カテゴリ2Unicode");
        String[] expected = new String[]{"カテゴリUñicode", "カテゴリ2Unicode"};
        Assert.assertArrayEquals(expected, test.getCategories());
    }

    @Test
    public void testJsonString() {
        test.addCategory("カテゴリUñicode");
        test.addCategory("カテゴリ2Unicode");
        //test.addCategory("𝄞");
        test.addCategory("鼖");
        String expected = "{\"category\":[\"\\u30ab\\u30c6\\u30b4\\u30eaU\\u00f1icode\",\"\\u30ab\\u30c6\\u30b4\\u30ea2Unicode\",\"\\ud87e\\ude1b\"]}";
        Assert.assertEquals(expected, test.jsonString());
    }

    @Test
    public void testRawJsonString() {
        test.addCategory("カテゴリUñicode");
        test.addCategory("カテゴリ2Unicode");
        test.addCategory("鼖");
        String expected = "{\"category\":[\"カテゴリUñicode\",\"カテゴリ2Unicode\",\"鼖\"]}";
        Assert.assertEquals(expected, test.rawJsonString());
    }

    @Test
    public void testAddCategories() throws JSONException {
        String[] expected = new String[]{"test", "test2"};
        test.addCategories(expected);
        Assert.assertArrayEquals(expected, test.getCategories());
    }

    @Test
    public void testAddSection() throws JSONException {
        test.addSection("test", "value");
        JSONObject testObject = test.getSections();
        JSONObject expected = new JSONObject().put("test", "value");
        JSONAssert.assertEquals(expected, testObject, false);
    }

    @Test
    public void testAddFilter() throws JSONException {
        test.addFilter("test", "setting", "value");
        String testObject = test.getFilters().toString();
        JSONObject expected = new JSONObject();
        expected.put("test", new JSONObject().put("settings", new JSONObject().put("setting", "value")));
        JSONAssert.assertEquals(expected.toString(), testObject, false);
    }

    @Test
    public void testSetASMGroupId() throws JSONException {
        test.setASMGroupId(1);
        Assert.assertEquals(new Integer(1), test.getASMGroupId());
    }

    @Test
    public void testSetSendAt() throws JSONException {
        int expected = 12345;
        test.setSendAt(expected);
        Assert.assertEquals(expected, test.getSendAt());
    }

    @Test
    public void testSetIpPool() throws JSONException {
        String expected = "transactional";
        test.setIpPool(expected);
        Assert.assertEquals(expected, test.getIpPool());
    }
}


src/main/java/com/revinate/sendgrid/smtpapi/SMTPAPI.java
package com.revinate.sendgrid.smtpapi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.*;

public class SMTPAPI {

    private static final String VERSION = "1.2.1";

    private static final String FIELD_ASM_GROUP_ID = "asm_group_id";
    private static final String FIELD_SEND_AT = "send_at";
    private static final String FIELD_IP_POOL = "ip_pool";
    private static final String FIELD_TO = "to";
    private static final String FIELD_CATEGORY = "category";
    private static final String FIELD_UNIQUE_ARGS = "unique_args";
    private static final String FIELD_SECTION = "section";
    private static final String FIELD_SUB = "sub";
    private static final String FIELD_FILTERS = "filters";
    private static final String FIELD_SETTINGS = "settings";
    private static final String EMAIL_FORMAT = "%s <%s>";

    private JSONObject header = new JSONObject();

    public String getVersion() {
        return VERSION;
    }

    public String toSmtpApiHeader() {
        return escapeUnicode(header.toString());
    }

    public String toRawSmtpApiHeader() {
        return header.toString();
    }

    public Integer getAsmGroupId() {
        return getInt(FIELD_ASM_GROUP_ID);
    }

    public SMTPAPI setAsmGroupId(Integer val) {
        header.put(FIELD_ASM_GROUP_ID, val);
        return this;
    }

    public Integer getSendAt() {
        return getInt(FIELD_SEND_AT);
    }

    public SMTPAPI setSendAt(Integer val) {
        header.put(FIELD_SEND_AT, val);
        return this;
    }

    public String getIpPool() {
        return header.optString(FIELD_IP_POOL, null);
    }

    public SMTPAPI setIpPool(String ipPool) {
        header.put(FIELD_IP_POOL, ipPool);
        return this;
    }

    public List<String> getSmtpApiTos() {
        return toListOfStrings(header.optJSONArray(FIELD_TO));
    }

    public SMTPAPI setSmtpApiTos(List<String> tos) {
        header.put(FIELD_TO, tos);
        return this;
    }

    public SMTPAPI addSmtpApiTo(String to) {
        header.append(FIELD_TO, to);
        return this;
    }

    public SMTPAPI addSmtpApiTo(String to, String name) {
        return addSmtpApiTo(String.format(EMAIL_FORMAT, name, to));
    }

    public List<String> getCategories() {
        return toListOfStrings(header.optJSONArray(FIELD_CATEGORY));
    }

    public SMTPAPI setCategories(List<String> categories) {
        header.put(FIELD_CATEGORY, categories);
        return this;
    }

    public SMTPAPI addCategory(String category) {
        header.append(FIELD_CATEGORY, category);
        return this;
    }

    public Map<String, String> getUniqueArgs() {
        return toMapOfStrings(header.optJSONObject(FIELD_UNIQUE_ARGS));
    }

    public SMTPAPI setUniqueArgs(Map<String, String> args) {
        header.put(FIELD_UNIQUE_ARGS, args);
        return this;
    }

    public String getUniqueArg(String key) {
        return getStringFromMap(FIELD_UNIQUE_ARGS, key);
    }

    public SMTPAPI setUniqueArg(String key, String val) {
        setInMap(FIELD_UNIQUE_ARGS, key, val);
        return this;
    }

    public Map<String, String> getSections() {
        return toMapOfStrings(header.optJSONObject(FIELD_SECTION));
    }

    public SMTPAPI setSections(Map<String, String> sections) {
        header.put(FIELD_SECTION, sections);
        return this;
    }

    public String getSection(String key) {
        return getStringFromMap(FIELD_SECTION, key);
    }

    public SMTPAPI setSection(String key, String val) {
        setInMap(FIELD_SECTION, key, val);
        return this;
    }

    public Map<String, List<String>> getSubstitutions() {
        return toMapOfListsOfStrings(header.optJSONObject(FIELD_SUB));
    }

    public SMTPAPI setSubstitutions(Map<String, List<String>> substitutions) {
        header.put(FIELD_SUB, substitutions);
        return this;
    }

    public List<String> getSubstitution(String key) {
        return getListOfStringsFromMap(FIELD_SUB, key);
    }

    public SMTPAPI setSubstitution(String key, List<String> vals) {
        setInMap(FIELD_SUB, key, vals);
        return this;
    }

    public SMTPAPI addValueToSubstitution(String key, String val) {
        appendToListInMap(FIELD_SUB, key, val);
        return this;
    }

    public Map<String, Map<String, Object>> getFilters() {
        Map<String, Map<String, Object>> filters = new HashMap<String, Map<String, Object>>();
        JSONObject json = header.optJSONObject(FIELD_FILTERS);
        if (json != null) {
            for (Object filterName : json.keySet()) {
                if (filterName instanceof String) {
                    String filterNameString = (String) filterName;
                    Map<String, Object> filter = toFilter(json.optJSONObject(filterNameString));
                    filters.put(filterNameString, filter);
                }
            }
        }
        return filters;
    }

    public SMTPAPI setFilters(Map<String, Map<String, Object>> filters) throws SMTPAPIException {
        Map<String, Map<String, Map<String, Object>>> filtersMap =
                new HashMap<String, Map<String, Map<String, Object>>>();
        for (Map.Entry<String, Map<String, Object>> filter : filters.entrySet()) {
            filtersMap.put(filter.getKey(), toFilterMap(filter.getValue()));
        }
        header.put(FIELD_FILTERS, filtersMap);
        return this;
    }

    public Map<String, Object> getFilter(String filterName) {
        JSONObject json = header.optJSONObject(FIELD_FILTERS);
        if (json != null) {
            return toFilter(json.optJSONObject(filterName));
        } else {
            return Collections.emptyMap();
        }
    }

    public SMTPAPI setFilter(String filterName, Map<String, Object> filter) throws SMTPAPIException {
        setInMap(FIELD_FILTERS, filterName, toFilterMap(filter));
        return this;
    }

    public SMTPAPI setSettingInFilter(String filterName, String settingName, Object settingVal) throws SMTPAPIException {
        if (!(settingVal instanceof Integer) && !(settingVal instanceof String)) {
            throw new SMTPAPIException("Filter setting value must be an integer or a String");
        }
        JSONObject filters = header.optJSONObject(FIELD_FILTERS);
        if (filters == null) {
            filters = new JSONObject();
        }
        JSONObject filter = filters.optJSONObject(filterName);
        if (filter == null) {
            filter = new JSONObject();
        }
        JSONObject settings = filter.optJSONObject(FIELD_SETTINGS);
        if (settings == null) {
            settings = new JSONObject();
        }
        settings.put(settingName, settingVal);
        filter.put(FIELD_SETTINGS, settings);
        filters.put(filterName, filter);
        header.put(FIELD_FILTERS, filters);
        return this;
    }

    private Integer getInt(String field) {
        try {
            return header.getInt(field);
        } catch (JSONException e) {
            return null;
        }
    }

    private String getStringFromMap(String field, String key) {
        JSONObject json = header.optJSONObject(field);
        return json == null? null : json.optString(key, null);
    }

    private List<String> getListOfStringsFromMap(String field, String key) {
        JSONObject json = header.optJSONObject(field);
        if (json != null) {
            return toListOfStrings(json.optJSONArray(key));
        }
        return Collections.emptyList();
    }

    private void setInMap(String field, String key, String val) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.put(key, val);
        header.put(field, json);
    }

    private void setInMap(String field, String key, List<String> vals) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.put(key, vals);
        header.put(field, json);
    }

    private void setInMap(String field, String key, Map<String, Map<String, Object>> val) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.put(key, val);
        header.put(field, json);
    }

    private void appendToListInMap(String field, String key, String val) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.append(key, val);
        header.put(field, json);
    }

    private List<String> toListOfStrings(JSONArray json) {
        List<String> parse = new ArrayList<String>();
        if (json != null) {
            for (int i = 0; i < json.length(); i++) {
                parse.add(json.optString(i, null));
            }
        }
        return parse;
    }

    private Map<String, Object> toMap(JSONObject json) {
        Map<String, Object> parse = new HashMap<String, Object>();
        if (json != null) {
            for (Object key : json.keySet()) {
                if (key instanceof String) {
                    String keyString = (String) key;
                    parse.put(keyString, json.opt(keyString));
                }
            }
        }
        return parse;
    }

    private Map<String, String> toMapOfStrings(JSONObject json) {
        Map<String, String> parse = new HashMap<String, String>();
        if (json != null) {
            for (Object key : json.keySet()) {
                if (key instanceof String) {
                    String keyString = (String) key;
                    parse.put(keyString, json.optString(keyString, null));
                }
            }
        }
        return parse;
    }

    private Map<String, List<String>> toMapOfListsOfStrings(JSONObject json) {
        Map<String, List<String>> parse = new HashMap<String, List<String>>();
        if (json != null) {
            for (Object key : json.keySet()) {
                if (key instanceof String) {
                    String keyString = (String) key;
                    JSONArray array = json.optJSONArray(keyString);
                    parse.put(keyString, toListOfStrings(array));
                }
            }
        }
        return parse;
    }

    private Map<String, Map<String, Object>> toFilterMap(Map<String, Object> filter) throws SMTPAPIException {
        Map<String, Map<String, Object>> filterMap = new HashMap<String, Map<String, Object>>();
        if (filter != null) {
            Map<String, Object> settingsMap = new HashMap<String, Object>();
            for (Map.Entry<String, Object> setting : filter.entrySet()) {
                Object settingVal = setting.getValue();
                if (!(settingVal instanceof Integer) && !(settingVal instanceof String)) {
                    throw new SMTPAPIException("Filter setting value must be an integer or a String");
                }
                settingsMap.put(setting.getKey(), settingVal);
            }
            filterMap.put(FIELD_SETTINGS, settingsMap);
        }
        return filterMap;
    }

    private Map<String, Object> toFilter(JSONObject json) {
        if (json != null) {
            return toMap(json.optJSONObject(FIELD_SETTINGS));
        } else {
            return Collections.emptyMap();
        }
    }

    private String escapeUnicode(String input) {
        StringBuilder sb = new StringBuilder();
        int[] codePoints = toCodePointArray(input);
        for (int codePoint : codePoints) {
            if (codePoint > 65535) {
                // surrogate pair
                int hi = (codePoint - 0x10000) / 0x400 + 0xD800;
                int lo = (codePoint - 0x10000) % 0x400 + 0xDC00;
                sb.append(String.format("\\u%04x\\u%04x", hi, lo));
            } else if (codePoint > 127) {
                sb.append(String.format("\\u%04x", codePoint));
            } else {
                sb.append(String.format("%c", codePoint));
            }
        }
        return sb.toString();
    }

    private int[] toCodePointArray(String input) {
        int len = input.length();
        int[] codePointArray = new int[input.codePointCount(0, len)];
        for (int i = 0, j = 0; i < len; i = input.offsetByCodePoints(i, 1)) {
            codePointArray[j++] = input.codePointAt(i);
        }
        return codePointArray;
    }
}


src/main/java/com/revinate/sendgrid/smtpapi/SmtpApiImpl.java
package com.revinate.sendgrid.smtpapi;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.*;

public class SmtpApiImpl implements SmtpApi {

    private static final String VERSION = "2.0.0";

    private static final String FIELD_ASM_GROUP_ID = "asm_group_id";
    private static final String FIELD_SEND_AT = "send_at";
    private static final String FIELD_IP_POOL = "ip_pool";
    private static final String FIELD_TO = "to";
    private static final String FIELD_CATEGORY = "category";
    private static final String FIELD_UNIQUE_ARGS = "unique_args";
    private static final String FIELD_SECTION = "section";
    private static final String FIELD_SUB = "sub";
    private static final String FIELD_FILTERS = "filters";
    private static final String FIELD_SETTINGS = "settings";
    private static final String EMAIL_FORMAT = "%s <%s>";

    private JSONObject header = new JSONObject();

    @Override
    public String getVersion() {
        return VERSION;
    }

    @Override
    public String toSmtpApiHeader() {
        return escapeUnicode(header.toString());
    }

    @Override
    public String toRawSmtpApiHeader() {
        return header.toString();
    }

    @Override
    public Integer getAsmGroupId() {
        return getInt(FIELD_ASM_GROUP_ID);
    }

    @Override
    public SmtpApi setAsmGroupId(Integer val) {
        header.put(FIELD_ASM_GROUP_ID, val);
        return this;
    }

    @Override
    public Integer getSendAt() {
        return getInt(FIELD_SEND_AT);
    }

    @Override
    public SmtpApi setSendAt(Integer val) {
        header.put(FIELD_SEND_AT, val);
        return this;
    }

    @Override
    public String getIpPool() {
        return header.optString(FIELD_IP_POOL, null);
    }

    @Override
    public SmtpApi setIpPool(String ipPool) {
        header.put(FIELD_IP_POOL, ipPool);
        return this;
    }

    @Override
    public List<String> getSmtpApiTos() {
        return toListOfStrings(header.optJSONArray(FIELD_TO));
    }

    @Override
    public SmtpApi setSmtpApiTos(List<String> tos) {
        header.put(FIELD_TO, tos);
        return this;
    }

    @Override
    public SmtpApi addSmtpApiTo(String to) {
        header.append(FIELD_TO, to);
        return this;
    }

    @Override
    public SmtpApi addSmtpApiTo(String to, String name) {
        return addSmtpApiTo(String.format(EMAIL_FORMAT, name, to));
    }

    @Override
    public List<String> getCategories() {
        return toListOfStrings(header.optJSONArray(FIELD_CATEGORY));
    }

    @Override
    public SmtpApi setCategories(List<String> categories) {
        header.put(FIELD_CATEGORY, categories);
        return this;
    }

    @Override
    public SmtpApi addCategory(String category) {
        header.append(FIELD_CATEGORY, category);
        return this;
    }

    @Override
    public Map<String, String> getUniqueArgs() {
        return toMapOfStrings(header.optJSONObject(FIELD_UNIQUE_ARGS));
    }

    @Override
    public SmtpApi setUniqueArgs(Map<String, String> args) {
        header.put(FIELD_UNIQUE_ARGS, args);
        return this;
    }

    @Override
    public String getUniqueArg(String key) {
        return getStringFromMap(FIELD_UNIQUE_ARGS, key);
    }

    @Override
    public SmtpApi setUniqueArg(String key, String val) {
        setInMap(FIELD_UNIQUE_ARGS, key, val);
        return this;
    }

    @Override
    public Map<String, String> getSections() {
        return toMapOfStrings(header.optJSONObject(FIELD_SECTION));
    }

    @Override
    public SmtpApi setSections(Map<String, String> sections) {
        header.put(FIELD_SECTION, sections);
        return this;
    }

    @Override
    public String getSection(String key) {
        return getStringFromMap(FIELD_SECTION, key);
    }

    @Override
    public SmtpApi setSection(String key, String val) {
        setInMap(FIELD_SECTION, key, val);
        return this;
    }

    @Override
    public Map<String, List<String>> getSubstitutions() {
        return toMapOfListsOfStrings(header.optJSONObject(FIELD_SUB));
    }

    @Override
    public SmtpApi setSubstitutions(Map<String, List<String>> substitutions) {
        header.put(FIELD_SUB, substitutions);
        return this;
    }

    @Override
    public List<String> getSubstitution(String key) {
        return getListOfStringsFromMap(FIELD_SUB, key);
    }

    @Override
    public SmtpApi setSubstitution(String key, List<String> vals) {
        setInMap(FIELD_SUB, key, vals);
        return this;
    }

    @Override
    public SmtpApi addValueToSubstitution(String key, String val) {
        appendToListInMap(FIELD_SUB, key, val);
        return this;
    }

    @Override
    public Map<String, Map<String, Object>> getFilters() {
        Map<String, Map<String, Object>> filters = new HashMap<String, Map<String, Object>>();
        JSONObject json = header.optJSONObject(FIELD_FILTERS);
        if (json != null) {
            for (Object filterName : json.keySet()) {
                if (filterName instanceof String) {
                    String filterNameString = (String) filterName;
                    Map<String, Object> filter = toFilter(json.optJSONObject(filterNameString));
                    filters.put(filterNameString, filter);
                }
            }
        }
        return filters;
    }

    @Override
    public SmtpApi setFilters(Map<String, Map<String, Object>> filters) throws SmtpApiException {
        Map<String, Map<String, Map<String, Object>>> filtersMap =
                new HashMap<String, Map<String, Map<String, Object>>>();
        for (Map.Entry<String, Map<String, Object>> filter : filters.entrySet()) {
            filtersMap.put(filter.getKey(), toFilterMap(filter.getValue()));
        }
        header.put(FIELD_FILTERS, filtersMap);
        return this;
    }

    @Override
    public Map<String, Object> getFilter(String filterName) {
        JSONObject json = header.optJSONObject(FIELD_FILTERS);
        if (json != null) {
            return toFilter(json.optJSONObject(filterName));
        } else {
            return Collections.emptyMap();
        }
    }

    @Override
    public SmtpApi setFilter(String filterName, Map<String, Object> filter) throws SmtpApiException {
        setInMap(FIELD_FILTERS, filterName, toFilterMap(filter));
        return this;
    }

    @Override
    public SmtpApi setSettingInFilter(String filterName, String settingName, Object settingVal) throws SmtpApiException {
        if (!(settingVal instanceof Integer) && !(settingVal instanceof String)) {
            throw new SmtpApiException("Filter setting value must be an integer or a String");
        }
        JSONObject filters = header.optJSONObject(FIELD_FILTERS);
        if (filters == null) {
            filters = new JSONObject();
        }
        JSONObject filter = filters.optJSONObject(filterName);
        if (filter == null) {
            filter = new JSONObject();
        }
        JSONObject settings = filter.optJSONObject(FIELD_SETTINGS);
        if (settings == null) {
            settings = new JSONObject();
        }
        settings.put(settingName, settingVal);
        filter.put(FIELD_SETTINGS, settings);
        filters.put(filterName, filter);
        header.put(FIELD_FILTERS, filters);
        return this;
    }

    private Integer getInt(String field) {
        try {
            return header.getInt(field);
        } catch (JSONException e) {
            return null;
        }
    }

    private String getStringFromMap(String field, String key) {
        JSONObject json = header.optJSONObject(field);
        return json == null? null : json.optString(key, null);
    }

    private List<String> getListOfStringsFromMap(String field, String key) {
        JSONObject json = header.optJSONObject(field);
        if (json != null) {
            return toListOfStrings(json.optJSONArray(key));
        }
        return Collections.emptyList();
    }

    private void setInMap(String field, String key, String val) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.put(key, val);
        header.put(field, json);
    }

    private void setInMap(String field, String key, List<String> vals) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.put(key, vals);
        header.put(field, json);
    }

    private void setInMap(String field, String key, Map<String, Map<String, Object>> val) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.put(key, val);
        header.put(field, json);
    }

    private void appendToListInMap(String field, String key, String val) {
        JSONObject json = header.optJSONObject(field);
        if (json == null) {
            json = new JSONObject();
        }
        json.append(key, val);
        header.put(field, json);
    }

    private List<String> toListOfStrings(JSONArray json) {
        List<String> parse = new ArrayList<String>();
        if (json != null) {
            for (int i = 0; i < json.length(); i++) {
                parse.add(json.optString(i, null));
            }
        }
        return parse;
    }

    private Map<String, Object> toMap(JSONObject json) {
        Map<String, Object> parse = new HashMap<String, Object>();
        if (json != null) {
            for (Object key : json.keySet()) {
                if (key instanceof String) {
                    String keyString = (String) key;
                    parse.put(keyString, json.opt(keyString));
                }
            }
        }
        return parse;
    }

    private Map<String, String> toMapOfStrings(JSONObject json) {
        Map<String, String> parse = new HashMap<String, String>();
        if (json != null) {
            for (Object key : json.keySet()) {
                if (key instanceof String) {
                    String keyString = (String) key;
                    parse.put(keyString, json.optString(keyString, null));
                }
            }
        }
        return parse;
    }

    private Map<String, List<String>> toMapOfListsOfStrings(JSONObject json) {
        Map<String, List<String>> parse = new HashMap<String, List<String>>();
        if (json != null) {
            for (Object key : json.keySet()) {
                if (key instanceof String) {
                    String keyString = (String) key;
                    JSONArray array = json.optJSONArray(keyString);
                    parse.put(keyString, toListOfStrings(array));
                }
            }
        }
        return parse;
    }

    private Map<String, Map<String, Object>> toFilterMap(Map<String, Object> filter) throws SmtpApiException {
        Map<String, Map<String, Object>> filterMap = new HashMap<String, Map<String, Object>>();
        if (filter != null) {
            Map<String, Object> settingsMap = new HashMap<String, Object>();
            for (Map.Entry<String, Object> setting : filter.entrySet()) {
                Object settingVal = setting.getValue();
                if (!(settingVal instanceof Integer) && !(settingVal instanceof String)) {
                    throw new SmtpApiException("Filter setting value must be an integer or a String");
                }
                settingsMap.put(setting.getKey(), settingVal);
            }
            filterMap.put(FIELD_SETTINGS, settingsMap);
        }
        return filterMap;
    }

    private Map<String, Object> toFilter(JSONObject json) {
        if (json != null) {
            return toMap(json.optJSONObject(FIELD_SETTINGS));
        } else {
            return Collections.emptyMap();
        }
    }

    private String escapeUnicode(String input) {
        StringBuilder sb = new StringBuilder();
        int[] codePoints = toCodePointArray(input);
        for (int codePoint : codePoints) {
            if (codePoint > 65535) {
                // surrogate pair
                int hi = (codePoint - 0x10000) / 0x400 + 0xD800;
                int lo = (codePoint - 0x10000) % 0x400 + 0xDC00;
                sb.append(String.format("\\u%04x\\u%04x", hi, lo));
            } else if (codePoint > 127) {
                sb.append(String.format("\\u%04x", codePoint));
            } else {
                sb.append(String.format("%c", codePoint));
            }
        }
        return sb.toString();
    }

    private int[] toCodePointArray(String input) {
        int len = input.length();
        int[] codePointArray = new int[input.codePointCount(0, len)];
        for (int i = 0, j = 0; i < len; i = input.offsetByCodePoints(i, 1)) {
            codePointArray[j++] = input.codePointAt(i);
        }
        return codePointArray;
    }
}
